internal/ingress/metric/collector/stats.go

```diff
@@ -0,0 +1,193 @@
+	listener             *net.UDPConn
+	ns                   string
+	watchClass           string
+	port                 int
+}
+
+// NewInstance creates a new StatsCollector instance
+func NewInstance(ns string, class string, port int) (*StatsCollector, error) {
+	sc := StatsCollector{}
+
+	listener, err := newUDPListener(port)
+
+	if err != nil {
+		return nil, err
+	}
+
+	sc.listener = listener
+	sc.ns = ns
+	sc.watchClass = class
+	sc.port = port
+
+	tags := []string{"host", "status", "remote_address", "remote_user", "protocol", "method", "path", "upstream_name", "upstream_ip", "upstream_status", "namespace", "ingress", "service"}
+
+	sc.upstreamResponseTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "upstream_response_time_seconds",
+			Help:      "The time spent on receiving the response from the upstream server",
+			Namespace: ns,
+		},
+		tags,
+	)
+
+	sc.requestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "request_duration_seconds",
+			Help:      "The request processing time in seconds",
+			Namespace: ns,
+		},
+		tags,
+	)
+
+	sc.requestLength = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "request_length_bytes",
+			Help:      "The request length (including request line, header, and request body)",
+			Namespace: ns,
+			Buckets:   prometheus.LinearBuckets(10, 10, 10), // 10 buckets, each 10 bytes wide.
+		},
+		tags,
+	)
+
+	sc.bytesSent = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "bytes_sent",
+			Help:      "The the number of bytes sent to a client",
+			Namespace: ns,
+			Buckets:   prometheus.ExponentialBuckets(10, 10, 7), // 7 buckets, exponential factor of 10.
+		},
+		tags,
+	)
+
+	prometheus.MustRegister(sc.upstreamResponseTime)
+	prometheus.MustRegister(sc.requestDuration)
+	prometheus.MustRegister(sc.requestLength)
+	prometheus.MustRegister(sc.bytesSent)
+	return &sc, nil
+}
+
+func (sc *StatsCollector) handleMessage(msg []byte) {
+	glog.V(5).Infof("msg: %v", string(msg))
+
+	// Unmarshall bytes
+	var stats udpData
+	err := json.Unmarshal(msg, &stats)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create Labels Map
+	labels := prometheus.Labels{
+		"host":            stats.Host,
+		"status":          stats.Status,
+		"remote_address":  stats.RemoteAddress,
+		"remote_user":     stats.RemoteUser,
+		"protocol":        stats.Protocol,
+		"method":          stats.Method,
+		"path":            stats.Path,
+		"upstream_name":   stats.UpstreamName,
+		"upstream_ip":     stats.UpstreamIP,
+		"upstream_status": stats.UpstreamStatus,
+		"namespace":       stats.Namespace,
+		"ingress":         stats.Ingress,
+		"service":         stats.Service,
+	}
+
+	// Emit metrics
+	if stats.UpstreamResponseTime != -1 {
+		upstreamResponseTimeMetric, err := sc.upstreamResponseTime.GetMetricWith(labels)
+		if err != nil {
+			glog.Errorf("Error fetching upstream response time metric: %v", err)
```

Comment:
Can you create a gauge metrics to show when was the last time that the collector ran without errors? You can use the same idea from [here](https://github.com/kubernetes/ingress-nginx/blob/master/internal/ingress/controller/metrics.go#L44-L57)

Reply:
Yup :)